src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/X509Tools.java

```diff
@@ -213,4 +213,54 @@ private static Object getX509Identity(final KeycloakSession session, final HttpR
         }
         return null;
     }
+
+    /**
+     * Get x509 subject DN from form context.
+     *
+     * @param context a Keycloak form context
+     * @return String
+     */
+    public static String getX509SubjectDN(final FormContext context) {
+        if (context.getSession() == null || context.getHttpRequest() == null || context.getSession().getProvider(X509ClientCertificateLookup.class) == null) {
+            return null;
+        }
+
+        return getX509SubjectDN(context.getSession(), context.getHttpRequest(), context.getSession().getProvider(X509ClientCertificateLookup.class));
+    }
+
+        /**
+     * Get x509 subject DN from form context.
+     *
+     * @param context a Keycloak form context
+     * @return String
+     */
+    public static String getX509SubjectDN(RequiredActionContext context) {
+        if (context.getSession() == null || context.getHttpRequest() == null || context.getSession().getProvider(X509ClientCertificateLookup.class) == null) {
+            return null;
+        }
+
+        return getX509SubjectDN(context.getSession(), context.getHttpRequest(), context.getSession().getProvider(X509ClientCertificateLookup.class));
+    }
+
+    /**
+     * Get x509 subject DN from form context.
+     *
+     * @param Keycloak form keycloak session
+     * @param Keycloak form context http request
+     * @param Keycloak form keycloak session provider certificate lookup
+     * @return String
+     */
+    public static String getX509SubjectDN(KeycloakSession session, HttpRequest httpRequest,
+            X509ClientCertificateLookup provider) {
+
+                try {
+                    X509Certificate[] certs = provider.getCertificateChain(httpRequest);
+                    return certs[0].getSubjectX500Principal().getName();
```

Comment:
Wrap this with `if (certs.length >= 0)` (and let the final `return null` run if empty)

Reply:
updated